#159965305 Create User Profiles

pngg opened the pull request:
#### What does this PR do?
Create user profiles

#### Description of Task to be completed
- Create a user profile when a user is created

#### How should this be manually tested?
- Run git clone `https://github.com/andela/ah-magnificent6.git`
- Git checkout `ft-create-user-profile-159965305`
- Make virtual environment with `python -m venv env`
- Activate the virtual environment with `. env/bin/activate`
- Install project requirements with `pip install -r requirements.txt`

```
Using Postman
Register new user: (POST) http://localhost:8000/api/users/signup/

payload
{
	"email": "joe3@company.com",
	"password": "pass1234",
	"username": "joe3"
}

Retrieve user with profile info: (GET) http://localhost:8000/api/user/{user_id}

payload
{
	"user":{
		"email": "joe3@company.com",
		"first_name": "Joe",
		"last_name": "Doe",
		"birth_date": "1970-03-04",
		"bio": "hI I am Joe",
		"city": "Nairobi",
		"country": "Kenya",
		"avatar":  "http://www.google.com",
		"phone": "21474836",
		"website": "http://www.google.com"
		}
}
```
#### What are the relevant pivotal tracker stories?
[Users should be able to create their profiles](https://www.pivotaltracker.com/story/show/159965305)

Commit:
[Feature #159965305] Add test for create profile

Commit:
[Feature #159965305] Create Profile model

Commit:
[Feature #159965305] Update Profile model

Commit:
[Feature #159965305] Create Serializer class for Profile

Commit:
[Feature #159965305] Create Renderer class for Profile

Commit:
[Feature #159965305] Add ProfileSerializer to UserSerializer

Commit:
[Feature #159965305] Update requirements.txt

Commit:
[Feature #159965305] Configure MEDIA_URL and MEDIA_ROOT

Commit:
[Feature #159965305] Update Profile model

Commit:
[Feature #159965305] Update UserSerializer

Commit:
[Feature #159965305] Update Profile model

Commit:
[Feature #159965305] Create shared TimeStamped model

Commit:
[Feature #159965305] Update serailizer_data for UserRetrieveUpdateAPIView

Commit:
[Feature #159965305] Add comments to the signal methods

Commit:
[Feature #159965305] Update Profile tests

Commit:
[Feature #159965305] Update UserSerializer

Commit:
[Feature #159965305] Update UserSerializer

Commit:
[Feature #159965305] Save user profile data

Commit:
[Feature #159965305] Update Profile model

Commit:
[Feature #159965305] Update ProfileSerializer

Commit:
[Feature #159965305] Update Profile tests

meshack-mbuvi commented on authors/apps/authentication/serializers.py:
```diff
@@ -190,6 +190,6 @@ def update(self, instance, validated_data):
         Save the user profile after update
         """
-        instance.profile.save()
+        # instance.profile.save()
```
You are supposed to delete unwanted code segments rather than commenting them out.

meshack-mbuvi commented on authors/apps/authentication/views.py:
```diff
@@ -127,6 +127,22 @@ def retrieve(self, request, *args, **kwargs):
 
+        serializer_data = {
+            'username': serializer_data.get('username', request.user.username),
+            'email': serializer_data.get('email', request.user.email),
```
Why do you set them one by one instead of passing an instance and request data to the serializer?For example, by having `serializer = serializer_class(user, serializer_data)?`

pngg replied:
@meshack-mbuvi I did that because I need to capture the `profile` key set in the UserSerializer. `i.e profile_data = validated_data.pop('profile')`

Commit:
[Feature #159965305] Auto-generate gravatar url from user email

Commit:
feat(profiles): create and update user profiles

- create user profile
- update user profile
- auto-generate avatar url

[Finishes #159965305]

parseen254 commented on requirements.txt:
```diff
@@ -23,8 +28,10 @@ PyJWT==1.4.2
 uritemplate==3.0.0
 urllib3==1.23
 whitenoise==4.0
-django-rest-swagger==2.2.0
```
@pngg we still need swagger.

pngg replied:
restored needed packages

parseen254 commented on requirements.txt:
```diff
@@ -23,8 +28,10 @@ PyJWT==1.4.2
 pytz==2018.5
 requests==2.19.1
 simplejson==3.16.0
+pylint==2.1.1
```
Let's be mindful not to add any unnecessary packages to requirements.txt

Commit:
[Feature #159965305] Update requirements.txt

parseen254 commented on authors/apps/articles/urls.py:
```diff
@@ -5,7 +5,7 @@
 urlpatterns = [
-    path('articles/', ArticleAPIView.as_view(), name='create'),
-    path('articles/<str:pk>', ArticleDetailsView.as_view(),
+    path('', ArticleAPIView.as_view(), name='create'),
```
@pngg Updating these would be going out of scope of your task. Create a chore for this separately on PT

pngg replied:
Noted

parseen254 commented:
Also make sure to work on swagger docs for this feature.

Commit:
[Feature #159965305] Update Profile tests

Bileonaire commented:
the user profile is working well. I have noticed that immediately I sign up, it reflects in the profiles table in the database.
